tools/c7n_traildb/trailes.py

```diff
@@ -0,0 +1,277 @@
+        es_kwargs['connection_class'] = MyConnection
+        es_kwargs['proxies'] = {'https': proxy}
+
+    return Elasticsearch(host, **es_kwargs)
+
+
+def index_events(client, events):
+    results = helpers.streaming_bulk(client, events)
+    for status, r in results:
+        if not status:
+            log.debug("index err result %s", r)
+
+
+def dict_factory(cursor, row):
+    """Returns a sqlite row factory that returns a dictionary"""
+    d = {}
+    for idx, col in enumerate(cursor.description):
+        d[col[0]] = row[idx]
+    return d
+
+
+def fetch_events(cursor, query, config, account_name):
+    """Generator that returns the events"""
+    for event in cursor.execute(query):
+        event['account'] = account_name
+        event['_index'] = config['indexer']['idx_name']
+        event['_type'] = config['indexer'].get('idx_type', 'traildb')
+        yield event
+
+
+def get_traildb(bucket, key, session_factory, directory):
+    local_db_file = directory + "/traildb" + \
+        str(thread.get_ident())
+    local_bz2_file = local_db_file + '.bz2'
+
+    s3 = local_session(session_factory).resource('s3')
+    s3.Bucket(bucket).download_file(key['Key'], local_bz2_file)
+
+    # Decompress the traildb file
+    # (use the system bunzip2 command because it's faster than the python bzip2 library)
+    subprocess.call(['bunzip2', '-f', local_bz2_file])
+
+    return local_db_file
+
+
+def valid_date(key, config_date):
+    """ traildb bucket folders are not zero-padded so this validation
+        checks that the keys returned by the paginator are
+        *really after* the config date
+    """
+    key_date = "/".join(key.split("/")[4:7])
+    return parse_date(key_date) > parse_date(config_date)
+
+
+def index_account_trails(config, account, region, date, directory):
+    es_client = get_es_client(config)
+
+    s3 = local_session(
+        lambda : assumed_session(account['role'], 'TrailIndex')).client('s3')
+
+    bucket = account['bucket']
+    key_prefix = "accounts/" + account['name'] + "/" + region + "/traildb"
+    marker = key_prefix + "/" + date + "/trail.db.bz2"
+
+    p = s3.get_paginator('list_objects_v2').paginate(
+        Bucket=bucket,
+        Prefix=key_prefix,
+        StartAfter=marker,
+    )
+
+    with ThreadPoolExecutor(max_workers=20) as w:
+        for key_set in p:
+            if 'Contents' not in key_set:
+                continue
+            keys = []
+            for k in key_set['Contents']:
+                if (k['Key'].endswith('trail.db.bz2') and valid_date(k['Key'], date)):
+                    keys.append(k)
+
+            futures = map(lambda k: w.submit(
+                get_traildb, bucket, k,
+                lambda : assumed_session(account['role'], 'TrailIndex'), directory),
+                keys)
+
+            for f in as_completed(futures):
+                local_db_file = f.result()
+                query = 'select * from events where user_agent glob \'*CloudCustodian*\''
```

Comment:
wondering if this should be configurable, the rest of the tools here are intended to be generic.